File: src/main/java/io/vproxy/vfx/ui/scene/VScene.java
package io.vproxy.vfx.ui.scene;

import io.vproxy.vfx.animation.AnimationGraph;
import io.vproxy.vfx.animation.AnimationGraphBuilder;
import io.vproxy.vfx.animation.AnimationNode;
import io.vproxy.vfx.control.scroll.VScrollPane;
import io.vproxy.vfx.control.scroll.NodeWithVScrollPane;
import io.vproxy.vfx.util.FXUtils;
import io.vproxy.vfx.util.algebradata.XYZTData;
import javafx.beans.value.ChangeListener;
import javafx.collections.ObservableList;
import javafx.scene.Node;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.Pane;
import javafx.scene.layout.Region;

import java.util.HashSet;
import java.util.Set;

public class VScene implements NodeWithVScrollPane {
    public static final int ANIMATION_DURATION_MILLIS = 300;
    private static final int SHOW_DELAY_MILLIS = 50;

    private final Pane root = new Pane();
    private final VScrollPane scrollPane = new VScrollPane();
    private final Pane content = new Pane();
    public final VSceneRole role;
    private ImageView backgroundImage;

    final AnimationNode<XYZTData> stateTop = new AnimationNode<>("top", new XYZTData(0, -1, 1, 0));
    final AnimationNode<XYZTData> stateRight = new AnimationNode<>("right", new XYZTData(1, 0, 1, 0));
    final AnimationNode<XYZTData> stateBottom = new AnimationNode<>("bottom", new XYZTData(0, 1, 1, 0));
    final AnimationNode<XYZTData> stateLeft = new AnimationNode<>("left", new XYZTData(-1, 0, 1, 0));
    final AnimationNode<XYZTData> stateCenterShown = new AnimationNode<>("center", new XYZTData(0, 0, 1, 1));
    final AnimationNode<XYZTData> stateFaded = new AnimationNode<>("faded", new XYZTData(0, 0, Float.MIN_VALUE * 2, 0));
    final AnimationNode<XYZTData> stateRemoved = new AnimationNode<>("removed", new XYZTData(0, 0, 1, 0),
        this::transferredToRemovedState);

    final AnimationGraph<XYZTData> animationGraph = new AnimationGraphBuilder<XYZTData>()
        .addNode(stateTop)
        .addNode(stateRight)
        .addNode(stateBottom)
        .addNode(stateLeft)
        .addNode(stateCenterShown)
        .addNode(stateFaded)
        .addNode(stateRemoved)
        .addEdge(stateRemoved, stateTop, SHOW_DELAY_MILLIS)
        .addEdge(stateRemoved, stateRight, SHOW_DELAY_MILLIS)
        .addEdge(stateRemoved, stateBottom, SHOW_DELAY_MILLIS)
        .addEdge(stateRemoved, stateLeft, SHOW_DELAY_MILLIS)
        .addEdge(stateRemoved, stateFaded, SHOW_DELAY_MILLIS)
        .addEdge(stateTop, stateRemoved, 1)
        .addEdge(stateRight, stateRemoved, 1)
        .addEdge(stateBottom, stateRemoved, 1)
        .addEdge(stateLeft, stateRemoved, 1)
        .addEdge(stateFaded, stateRemoved, 1)
        .addEdge(stateTop, stateCenterShown, ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
        .addEdge(stateRight, stateCenterShown, ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
        .addEdge(stateBottom, stateCenterShown, ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
        .addEdge(stateLeft, stateCenterShown, ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
        .addEdge(stateFaded, stateCenterShown, ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
        .addEdge(stateCenterShown, stateTop, ANIMATION_DURATION_MILLIS - 1)
        .addEdge(stateCenterShown, stateRight, ANIMATION_DURATION_MILLIS - 1)
        .addEdge(stateCenterShown, stateBottom, ANIMATION_DURATION_MILLIS - 1)
        .addEdge(stateCenterShown, stateLeft, ANIMATION_DURATION_MILLIS - 1)
        .addEdge(stateCenterShown, stateFaded, ANIMATION_DURATION_MILLIS - 1)
        .setStateTransferBeginCallback(this::animationStateTransferBegin)
        .setApply(this::applyAnimation)
        .build(stateRemoved);

    VSceneGroup.ProgressInformer progressInformer;
    ObservableList<Node> parentChildren;
    double xN;
    double yN;
    double x0;
    double y0;
    double x1;
    double y1;
    final Set<ChangeListener<? super Number>> changeListeners = new HashSet<>();
    Node bundle;
    Pane cover;
    VSceneHideMethod defaultHideMethod;

    public VScene(VSceneRole role) {
        this.role = role;
        scrollPane.setContent(content);
        FXUtils.observeWidthHeight(root, scrollPane.getNode());
        root.getChildren().add(scrollPane.getNode());

        root.widthProperty().addListener(ob -> updateBackgroundImagePos());
        root.heightProperty().addListener(ob -> updateBackgroundImagePos());
    }

    public Region getNode() {
        return root;
    }

    @Override
    public VScrollPane getScrollPane() {
        return scrollPane;
    }

    @Override
    public Region getSelfNode() {
        return getNode();
    }

    public Pane getContentPane() {
        return content;
    }

    public void enableAutoContentWidthHeight() {
        enableAutoContentWidth();
        enableAutoContentHeight();
    }

    public void enableAutoContentWidth() {
        FXUtils.observeWidth(root, content, -1);
    }

    public void enableAutoContentHeight() {
        FXUtils.observeHeight(root, content, -1);
    }

    private void transferredToRemovedState(AnimationNode<XYZTData> from, AnimationNode<XYZTData> to) {
        if (parentChildren != null && bundle != null) {
            parentChildren.remove(bundle);
        }
    }

    private void applyAnimation(AnimationNode<XYZTData> from, AnimationNode<XYZTData> to, XYZTData data) {
        if (from == stateRemoved || to == stateRemoved) {
            return; // do not animate in this state
        }

        double x = data.x;
        double y = data.y;
        double opacity = data.z;
        if (x < 0) {
            x = (xN - x0) * (-x) + x0;
        } else if (x > 0) {
            x = (x1 - x0) * x + x0;
        } else {
            x = x0;
        }
        if (y < 0) {
            y = (yN - y0) * (-y) + y0;
        } else if (y > 0) {
            y = (y1 - y0) * y + y0;
        } else {
            y = y0;
        }
        getNode().setLayoutX(x);
        getNode().setLayoutY(y);
        getNode().setOpacity(opacity);
        if (progressInformer != null) // it is null when constructing
            progressInformer.informUpdate(this, data.t);
    }

    private void animationStateTransferBegin(AnimationNode<XYZTData> from, AnimationNode<XYZTData> to) {
        if (from == stateRemoved) {
            if (bundle != null) {
                bundle.setOpacity(Float.MIN_VALUE * 2);
                bundle.setMouseTransparent(true);
                if (parentChildren != null) {
                    if (!parentChildren.contains(bundle)) {
                        parentChildren.add(bundle);
                    }
                }
            }
        } else if (from == stateTop || from == stateBottom || from == stateLeft || from == stateRight || from == stateFaded) {
            if (to == stateCenterShown) {
                if (bundle != null) {
                    bundle.setOpacity(1);
                    bundle.setMouseTransparent(false);
                }
            }
        }
    }

    public void setBackgroundImage(Image img) {
        if (img == null) {
            if (backgroundImage == null) {
                return;
            }
            root.getChildren().remove(backgroundImage);
            backgroundImage = null;
            return;
        }
        if (backgroundImage == null) {
            backgroundImage = new ImageView(img);
            backgroundImage.setPreserveRatio(true);
            root.getChildren().add(0, backgroundImage);
        } else {
            backgroundImage.setImage(img);
        }
        updateBackgroundImagePos();
    }

    private void updateBackgroundImagePos() {
        if (backgroundImage == null)
            return;
        if (backgroundImage.getImage() == null)
            return;
        var iw = backgroundImage.getImage().getWidth();
        var ih = backgroundImage.getImage().getHeight();
        var nw = root.getWidth();
        var nh = root.getHeight();
        if (nw <= 0 || nh <= 0 || iw <= 0 || ih <= 0)
            return;
        if (iw / ih > nw / nh) {
            backgroundImage.setFitHeight(nh);
            backgroundImage.setFitWidth(nh * iw / ih);
        } else {
            backgroundImage.setFitWidth(nw);
            backgroundImage.setFitHeight(nw * ih / iw);
        }
        var bw = backgroundImage.getFitWidth();
        var bh = backgroundImage.getFitHeight();
        backgroundImage.setLayoutX((nw - bw) / 2);
        backgroundImage.setLayoutY((nh - bh) / 2);
    }

    @SuppressWarnings("RedundantThrows")
    protected void beforeShowing() throws Exception {
    }

    protected void onHidden() {
    }
}

File: src-ts/src/ui/scene/VScene.ts
import { Pane } from '../../javafx/Pane.js';
import { Node } from '../../javafx/Node.js';
import type { ChangeListenerNumber } from '../../util/FXUtils.js';
import { FXUtils } from '../../util/FXUtils.js';
import { VScrollPane } from '../../control/scroll/VScrollPane.js';
import { AnimationGraphBuilder } from '../../animation/AnimationGraphBuilder.js';
import { AnimationNode } from '../../animation/AnimationNode.js';
import { AnimationGraph } from '../../animation/AnimationGraph.js';
import { XYZTData } from '../../util/algebradata/XYZTData.js';
import { VSceneRole } from './VSceneRole.js';
import { VSceneHideMethod } from './VSceneHideMethod.js';
import type { FXImage } from '../../javafx/ImageView.js';
import { ImageView } from '../../javafx/ImageView.js';
import type { NodeWithVScrollPane } from '../../control/scroll/NodeWithVScrollPane.js';

const SHOW_DELAY_MILLIS = 50;
const FADE_MIN = 1e-44; // Float.MIN_VALUE * 2 in Java; arbitrarily small positive.

export interface ProgressInformer {
  informUpdate(scene: VScene, p: number): void;
}

export class VScene implements NodeWithVScrollPane {
  static readonly ANIMATION_DURATION_MILLIS = 300;

  private readonly root: Pane;
  private readonly scrollPane: VScrollPane;
  private readonly content: Pane;
  readonly role: VSceneRole;
  private backgroundImage: ImageView | null = null;

  // Animation states — package-private for VSceneGroup access.
  readonly stateTop = new AnimationNode<XYZTData>('top', new XYZTData(0, -1, 1, 0));
  readonly stateRight = new AnimationNode<XYZTData>('right', new XYZTData(1, 0, 1, 0));
  readonly stateBottom = new AnimationNode<XYZTData>('bottom', new XYZTData(0, 1, 1, 0));
  readonly stateLeft = new AnimationNode<XYZTData>('left', new XYZTData(-1, 0, 1, 0));
  readonly stateCenterShown = new AnimationNode<XYZTData>('center', new XYZTData(0, 0, 1, 1));
  readonly stateFaded = new AnimationNode<XYZTData>('faded', new XYZTData(0, 0, FADE_MIN, 0));
  readonly stateRemoved = new AnimationNode<XYZTData>('removed', new XYZTData(0, 0, 1, 0),
    (_f, _t) => this.transferredToRemovedState());

  readonly animationGraph: AnimationGraph<XYZTData>;

  // Package-private state shared with VSceneGroup.
  progressInformer: ProgressInformer | null = null;
  parentChildren: import('../../javafx/Parent.js').ObservableList<Node> | null = null;
  xN = 0;
  yN = 0;
  x0 = 0;
  y0 = 0;
  x1 = 0;
  y1 = 0;
  changeListeners = new Set<ChangeListenerNumber>();
  bundle: Node | null = null;
  cover: Pane | null = null;
  defaultHideMethod: VSceneHideMethod | null = null;

  constructor(role: VSceneRole) {
    this.role = role;
    this.root = new Pane();
    this.scrollPane = new VScrollPane();
    this.content = new Pane();
    this.scrollPane.setContent(this.content);
    FXUtils.observeWidthHeight(this.root, this.scrollPane.getNode());
    this.root.getChildren().add(this.scrollPane.getNode());

    this.root.widthProperty.addListener(() => this.updateBackgroundImagePos());
    this.root.heightProperty.addListener(() => this.updateBackgroundImagePos());

    this.animationGraph = new AnimationGraphBuilder<XYZTData>()
      .addNode(this.stateTop)
      .addNode(this.stateRight)
      .addNode(this.stateBottom)
      .addNode(this.stateLeft)
      .addNode(this.stateCenterShown)
      .addNode(this.stateFaded)
      .addNode(this.stateRemoved)
      .addEdge(this.stateRemoved, this.stateTop, SHOW_DELAY_MILLIS)
      .addEdge(this.stateRemoved, this.stateRight, SHOW_DELAY_MILLIS)
      .addEdge(this.stateRemoved, this.stateBottom, SHOW_DELAY_MILLIS)
      .addEdge(this.stateRemoved, this.stateLeft, SHOW_DELAY_MILLIS)
      .addEdge(this.stateRemoved, this.stateFaded, SHOW_DELAY_MILLIS)
      .addEdge(this.stateTop, this.stateRemoved, 1)
      .addEdge(this.stateRight, this.stateRemoved, 1)
      .addEdge(this.stateBottom, this.stateRemoved, 1)
      .addEdge(this.stateLeft, this.stateRemoved, 1)
      .addEdge(this.stateFaded, this.stateRemoved, 1)
      .addEdge(this.stateTop, this.stateCenterShown, VScene.ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
      .addEdge(this.stateRight, this.stateCenterShown, VScene.ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
      .addEdge(this.stateBottom, this.stateCenterShown, VScene.ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
      .addEdge(this.stateLeft, this.stateCenterShown, VScene.ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
      .addEdge(this.stateFaded, this.stateCenterShown, VScene.ANIMATION_DURATION_MILLIS - SHOW_DELAY_MILLIS)
      .addEdge(this.stateCenterShown, this.stateTop, VScene.ANIMATION_DURATION_MILLIS - 1)
      .addEdge(this.stateCenterShown, this.stateRight, VScene.ANIMATION_DURATION_MILLIS - 1)
      .addEdge(this.stateCenterShown, this.stateBottom, VScene.ANIMATION_DURATION_MILLIS - 1)
      .addEdge(this.stateCenterShown, this.stateLeft, VScene.ANIMATION_DURATION_MILLIS - 1)
      .addEdge(this.stateCenterShown, this.stateFaded, VScene.ANIMATION_DURATION_MILLIS - 1)
      .setStateTransferBeginCallback((from, to) => this.animationStateTransferBegin(from, to))
      .setApply((from, to, data) => this.applyAnimation(from, to, data))
      .build(this.stateRemoved);
  }

  getNode(): Pane { return this.root; }
  getScrollPane(): VScrollPane { return this.scrollPane; }
  getSelfNode(): Pane { return this.getNode(); }
  getContentPane(): Pane { return this.content; }

  enableAutoContentWidthHeight(): void {
    this.enableAutoContentWidth();
    this.enableAutoContentHeight();
  }
  enableAutoContentWidth(): void {
    FXUtils.observeWidth(this.root, this.content, -1);
  }
  enableAutoContentHeight(): void {
    FXUtils.observeHeight(this.root, this.content, -1);
  }

  private transferredToRemovedState(): void {
    if (this.parentChildren !== null && this.bundle !== null) {
      this.parentChildren.remove(this.bundle);
    }
  }

  private applyAnimation(
    from: AnimationNode<XYZTData> | null,
    to: AnimationNode<XYZTData>,
    data: XYZTData,
  ): void {
    if (from === this.stateRemoved || to === this.stateRemoved) {
      return;
    }
    let x = data.x;
    let y = data.y;
    const opacity = data.z;
    if (x < 0) {
      x = (this.xN - this.x0) * (-x) + this.x0;
    } else if (x > 0) {
      x = (this.x1 - this.x0) * x + this.x0;
    } else {
      x = this.x0;
    }
    if (y < 0) {
      y = (this.yN - this.y0) * (-y) + this.y0;
    } else if (y > 0) {
      y = (this.y1 - this.y0) * y + this.y0;
    } else {
      y = this.y0;
    }
    this.getNode().setLayoutX(x);
    this.getNode().setLayoutY(y);
    this.getNode().setOpacity(opacity);
    if (this.progressInformer !== null) {
      this.progressInformer.informUpdate(this, data.t);
    }
  }

  private animationStateTransferBegin(
    from: AnimationNode<XYZTData>,
    to: AnimationNode<XYZTData>,
  ): void {
    if (from === this.stateRemoved) {
      if (this.bundle !== null) {
        this.bundle.setOpacity(FADE_MIN);
        this.bundle.setMouseTransparent(true);
        if (this.parentChildren !== null && !this.parentChildren.contains(this.bundle)) {
          this.parentChildren.add(this.bundle);
        }
      }
    } else if (
      from === this.stateTop ||
      from === this.stateBottom ||
      from === this.stateLeft ||
      from === this.stateRight ||
      from === this.stateFaded
    ) {
      if (to === this.stateCenterShown && this.bundle !== null) {
        this.bundle.setOpacity(1);
        this.bundle.setMouseTransparent(false);
      }
    }
  }

  setBackgroundImage(img: FXImage | null): void {
    if (img === null) {
      if (this.backgroundImage === null) return;
      this.root.getChildren().remove(this.backgroundImage);
      this.backgroundImage = null;
      return;
    }
    if (this.backgroundImage === null) {
      this.backgroundImage = new ImageView(img);
      this.backgroundImage.setPreserveRatio(true);
      // Insert at the back so content draws on top.
      this.root.getChildren().insert(0, this.backgroundImage);
    } else {
      this.backgroundImage.setImage(img);
    }
    this.updateBackgroundImagePos();
  }

  private updateBackgroundImagePos(): void {
    if (this.backgroundImage === null) return;
    const img = this.backgroundImage.getImage();
    if (img === null) return;
    const iw = img.width;
    const ih = img.height;
    const nw = this.root.getWidth();
    const nh = this.root.getHeight();
    if (nw <= 0 || nh <= 0 || iw <= 0 || ih <= 0) return;
    if (iw / ih > nw / nh) {
      this.backgroundImage.setFitHeight(nh);
      this.backgroundImage.setFitWidth((nh * iw) / ih);
    } else {
      this.backgroundImage.setFitWidth(nw);
      this.backgroundImage.setFitHeight((nw * ih) / iw);
    }
    const bw = this.backgroundImage.getFitWidth();
    const bh = this.backgroundImage.getFitHeight();
    this.backgroundImage.setLayoutX((nw - bw) / 2);
    this.backgroundImage.setLayoutY((nh - bh) / 2);
  }

  // In Java these are `protected` and called by VSceneGroup in the same
  // package. TypeScript has no package-private visibility, so we expose them
  // publicly — subclasses still override them freely.
  checkBeforeShowing(): boolean { return true; }
  beforeShowing(): void {}
  onShown(): void {}
  checkBeforeHiding(): boolean { return true; }
  beforeHiding(): void {}
  onHidden(): void {}
}

